src/general/interpolation.jl

```diff
@@ -565,42 +570,98 @@ end
             computed_density[point] = NaN
             neighbor_count[point] = 0
 
-            foreach(cache) do field
-                if field isa AbstractVector
-                    field[point] = NaN
-                else
-                    field[:, point] .= NaN
-                end
-            end
+            set_nan!(cache, point, ref_system)
         else
             # Normalize all quantities by the shepard coefficient
-            foreach(cache) do field
-                if field isa AbstractVector
-                    field[point] /= shepard_coefficient[point]
-                else
-                    field[:, point] ./= shepard_coefficient[point]
-                end
-            end
+            divide_by_shepard_coefficient!(cache, point, shepard_coefficient,
+                                           ref_system)
         end
     end
 
-    return (; computed_density=computed_density, neighbor_count, point_coords, cache...)
+    cache_ = map(x -> Array(x), cache)
+
+    return (; computed_density=Array(computed_density), point_coords=Array(point_coords),
+            neighbor_count=Array(neighbor_count), cache_...)
```

Comment:
Why do you move everything to the CPU here?

Reply:
I take that back. It's confusing that interpolation returns a CPU array when the whole simulation ran on the GPU and all the arrays I put into the interpolation are GPU arrays.